library/lua/gui.lua

```diff
@@ -951,8 +951,28 @@ BOLD_FRAME = FRAME_BOLD
 INTERIOR_FRAME = FRAME_INTERIOR
 INTERIOR_MEDIUM_FRAME = FRAME_INTERIOR_MEDIUM
 
+-- for compatibility with dynamic textures
+local function choose_frame_style(style)
+    if style == FRAME_WINDOW or style == WINDOW_FRAME then return make_frame('Window', true) end
+    if style == FRAME_PANEL or style == GREY_LINE_FRAME or style == PANEL_FRAME then
+        return make_frame('Panel', true)
+    end
+    if style == FRAME_MEDIUM or style == MEDIUM_FRAME then return make_frame('Medium', true) end
+    if style == FRAME_BOLD then return make_frame('Bold', true) end
+    if style == FRAME_INTERIOR or style == INTERIOR_FRAME then
+        local frame = make_frame('Thin', true)
+        frame.signature_pen = false
+        return frame
+    end
+    if style == FRAME_INTERIOR_MEDIUM or style == INTERIOR_MEDIUM_FRAME then
+        local frame = make_frame('Medium', true)
+        frame.signature_pen = false
+        return frame
+    end
+end
```

Comment:
This will only work for non-custom frames. What about if you replace the `tile` value with a function that returns a texpos? Then you can call that function in LuaApi.cpp in `decode_pen()` (replace the line `get_int_field(L, &pen.tile, idx, "tile", 0);`)

that is, extend Pen abstraction to allow the `tile` field to be either an integer or a function that returns an integer. If it's the former, then the current handling in `decode_pen()` stays the same. If it's a function, then call it to get the integer.

Reply:
This is just a crutch to bring current codebase to work with introduced pipeline. Should be reworked fro sure!
I'm not sure i understand you clearly:) Imho, cpp should know nothing about special cases.
Maybe it will be better to rework lua part to work with callbacks but not with static values? (more work, i guess)